openid-4-verifiable-credential-issuance-1_0.md

```diff
@@ -1119,6 +1121,67 @@ Cache-Control: no-store
 }
 ```
 
+# Callback Endpoint {#callback_endpoint}
+
+This endpoint is used to receive notification from the Wallet whether credential has been successfully received in the Wallet or not. The Credential Issuer needs to return `callback_id` in the Credential Response or a Batch Credential Response for the Wallet to be able to use this Endpoint. Support for this endpoint is OPTIONAL.
+
+This endpoint can be used after the Credential Issuer has sent Credential Response or Batch Credential Response. It enables the Credential Issuer to take subsequent actions after issuance, depending on whether the credential has been accepted and successully stored by the Wallet, rejected by the Wallet, or errors and other unforeseen circumstances have occurred during the Wallet's processing.
+
+Communication with the Callback Endpoint MUST utilize TLS.
+
+## Callback from the Wallet {#callback}
+
+The Wallet sends an HTTP POST request to the Callback Endpoint with the following parameters in the entity-body and using the `application/json` media type.
+
+* `issuer_state`: REQUIRED if the `issuer_state` parameter was present in the Credential Offer (see (#credential_offer)).
+* `credentials`: A JSON array of objects, where each object consists of the following parameters:
+  * `callback_id`: REQUIED. A JSON string received in Credential Response or Batch Credential Response.
+  * `status`: REQUIRED. Status whether the credential issuance was successful or not. The value MUST be either `success` or `error`.
+  * `error_description`: OPTIONAL. Human-readable ASCII [@!USASCII] text providing additional information, used to assist the Credential Issuer developer in understanding the error that occurred. Values for the `error_description`` parameter MUST NOT include characters outside the set %x20-21 / %x23-5B / %x5D-7E.
+
+Below is a non-normative example of a callback when credential issuance was successful:
+
+```
+POST /callback HTTP/1.1
+Host: server.example.com
+Content-Type: application/json
+{
+  "issuer_state": "eyJhbGciOiJSU0Et...FYUaBy",
+  "credentials": [
+    {
+    "callback_id": "3fwe98js",
+    "status": "success"
+    }
+  ]
+}
+```
+
+Below is a non-normative example of a callback when credential issuance was unsuccessful:
+
+```
+POST /callback HTTP/1.1
+Host: server.example.com
+Content-Type: application/json
+{
+  "issuer_state": "eyJhbGciOiJSU0Et...FYUaBy",
+  "credentials": [
+    {
+    "callback_id": "3fwe98js",
+    "status": "error",
+    "error_description": "..."
+    }
+  ]
+}
+```
+
+## Response from the Credential Issuer
+The Credential Issuer MUST respond with the HTTP status code 200 (OK) and use the `text/plain` media type:
+```
+HTTP/1.1 200 OK
+Content-Type: text/plain
```

Comment:
Would it be appropriate to use a 204 No Content?

Reply:
It makes sense

Reply:
eeee i think usually it would be 200 with `OK` or `ok` or something like that. so instead of changing to 204, would prefer reverting to `ok` or something and say that the content on the message can be anything

Reply:
I suggest something along the lines of:
- A wallet MUST interpret a response with a 2xx status code as meaning that the request was correctly processed by the issuer and no retry is needed. The wallet should also ignore any response payload if the status code is a 2xx.
- An issuer SHOULD return a 204 No Content however usage of 200 OK with an empty `text/plain` is also allowed.

IMO, a 204 fits well into what is being done:

> The 204 (No Content) status code indicates that the server has successfully fulfilled the request and that there is no additional content to send in the response content. 
In https://httpwg.org/specs/rfc9110.html#status.204

Reply:
Do we need to account for a response to an invalid call to the endpoint, i.e. when a request is made without a valid access token, or without a valid callback_id with an unauthorised or bad_request, or should the endpoint _always_ return 204?

i.e. do we need to update the language to say "_In response to a valid request_, the Credential Issuer MUST respond with the HTTP status code 204 (No Content) and use the `text/plain` media type."